lapce-app/src/editor/view.rs

```diff
@@ -843,6 +843,210 @@ impl EditorView {
+    fn paint_char_highlights(
+        &self,
+        cx: &mut PaintCx,
+        screen_lines: &ScreenLines,
+        highlight_line_cols: impl Iterator<Item = (usize, usize)>,
+    ) {
+        let view = &self.editor.view;
+        let config = self.editor.common.config.get_untracked();
+        let line_height = config.editor.line_height() as f64;
+
+        for (line, col) in highlight_line_cols {
+            // Is the given line on screen?
+            if let Some(line_info) = screen_lines.info.get(&line) {
+                let x0 = Self::calculate_col_x(view, line, col, true);
+                let x1 = Self::calculate_col_x(view, line, col + 1, false);
+
+                let y0 = line_info.y as f64;
+                let y1 = y0 + line_height;
+
+                let rect = Rect::new(x0, y0, x1, y1);
+
+                cx.stroke(
+                    &rect,
+                    config.get_color(LapceColor::EDITOR_FOREGROUND),
+                    1.0,
+                );
+            }
+        }
+    }
+
+    /// Paint scope lines between `(start_line, start_col)` and `(end_line, end_col)`.
+    fn paint_scope_lines(
+        &self,
+        cx: &mut PaintCx,
+        viewport: Rect,
+        screen_lines: &ScreenLines,
+        (start_line, start_col): (usize, usize),
+        (end_line, end_col): (usize, usize),
+    ) {
+        let view = &self.editor.view;
+        let doc = view.doc.get_untracked();
+        let config = self.editor.common.config.get_untracked();
+        let line_height = config.editor.line_height() as f64;
+        let brush = config.get_color(LapceColor::EDITOR_FOREGROUND);
+
+        if start_line == end_line {
+            if let Some(line_info) = screen_lines.info.get(&start_line) {
+                let x0 =
+                    Self::calculate_col_x(view, start_line, start_col + 1, false);
+                let x1 = Self::calculate_col_x(view, end_line, end_col, true);
+
+                if x0 < x1 {
+                    let y = line_info.y as f64 + line_height;
+
+                    let p0 = Point::new(x0, y);
+                    let p1 = Point::new(x1, y);
+                    let line = Line::new(p0, p1);
+
+                    cx.stroke(&line, brush, 1.0);
+                }
+            }
+        } else {
+            // Are start_line and end_line on screen?
+            let start_line_y = screen_lines
+                .info
+                .get(&start_line)
+                .map(|line_info| line_info.y as f64 + line_height);
+            let end_line_y = screen_lines
+                .info
+                .get(&end_line)
+                .map(|line_info| line_info.y as f64 + line_height);
+
+            // We only need to draw anything if start_line is on or before the visible section and
+            // end_line is on or after the visible section.
+            let y0 = start_line_y.or_else(|| {
+                (start_line < *screen_lines.lines.first().unwrap())
```

Comment:
Is it valid to assume that `screen_lines.lines` is non-empty and sorted?

Reply:
It's sorted but could be empty I think.

Reply:
I've changed it to not assume that `screen_lines.lines` is non-empty.